api/provider/aws/test_aws/task_list_test.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestTaskList(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockAWS := awsc.NewMockClient(ctrl)
+	tagStore := tag.NewMemoryStore()
+	mockConfig := mock_config.NewMockAPIConfig(ctrl)
+
+	mockConfig.EXPECT().Instance().Return("test").AnyTimes()
+
+	tags := models.Tags{
+		{
+			EntityID:   "tsk_id1",
+			EntityType: "task",
+			Key:        "name",
+			Value:      "tsk_name1",
+		},
+		{
+			EntityID:   "tsk_id1",
+			EntityType: "task",
+			Key:        "environment_id",
+			Value:      "env_id1",
+		},
+		{
+			EntityID:   "tsk_id1",
+			EntityType: "task",
+			Key:        "arn",
+			Value:      "arn:aws:ecs:region:012345678910:task/arn1",
+		},
+		{
+			EntityID:   "tsk_id2",
+			EntityType: "task",
+			Key:        "name",
+			Value:      "tsk_name2",
+		},
+		{
+			EntityID:   "tsk_id2",
+			EntityType: "task",
+			Key:        "environment_id",
+			Value:      "env_id2",
+		},
+		{
+			EntityID:   "tsk_id2",
+			EntityType: "task",
+			Key:        "arn",
+			Value:      "arn:aws:ecs:region:012345678910:task/arn2",
+		},
+		{
+			EntityID:   "env_id1",
+			EntityType: "environment",
+			Key:        "name",
+			Value:      "env_name1",
+		},
+		{
+			EntityID:   "env_id2",
+			EntityType: "environment",
+			Key:        "name",
+			Value:      "env_name2",
+		},
+	}
+
+	for _, tag := range tags {
+		if err := tagStore.Insert(tag); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	listClusterPagesFN := func(input *ecs.ListClustersInput, fn func(output *ecs.ListClustersOutput, lastPage bool) bool) error {
+		clusterARNs := []*string{
+			aws.String("arn:aws:ecs:region:012345678910:cluster/l0-test-env_id1"),
+			aws.String("arn:aws:ecs:region:012345678910:cluster/l0-test-env_id2"),
+		}
+
+		output := &ecs.ListClustersOutput{}
+		output.SetClusterArns(clusterARNs)
+
+		fn(output, true)
+
+		return nil
+	}
+
+	mockAWS.ECS.EXPECT().
+		ListClustersPages(&ecs.ListClustersInput{}, gomock.Any()).
+		Do(listClusterPagesFN).
+		Return(nil)
+
+	newListTaskPagesFN := func(taskARN int) func(input *ecs.ListTasksInput, fn func(output *ecs.ListTasksOutput, lastPage bool) bool) error {
+		listTaskPagesFN := func(input *ecs.ListTasksInput, fn func(output *ecs.ListTasksOutput, lastPage bool) bool) error {
+			arn := []*string{aws.String(fmt.Sprintf("arn:aws:ecs:region:012345678910:task/arn%d", taskARN))}
+
+			output := &ecs.ListTasksOutput{}
+			output.SetTaskArns(arn)
+			fn(output, true)
+
+			return nil
+		}
+
+		return listTaskPagesFN
+	}
+
+	for i, environmentID := range []string{"l0-test-env_id1", "l0-test-env_id2"} {
+		for j, status := range []string{ecs.DesiredStatusRunning, ecs.DesiredStatusStopped} {
+			listTasksInput := &ecs.ListTasksInput{}
+			listTasksInput.SetCluster(environmentID)
+			listTasksInput.SetDesiredStatus(status)
+			listTasksInput.SetStartedBy("test")
+
+			mockAWS.ECS.EXPECT().
+				ListTasksPages(listTasksInput, gomock.Any()).
+				Do(newListTaskPagesFN(i + j)).
```

Comment:
doesn't this create a `aws:arn.../arn0` on the first iteration? Seems like we should be testing that 4 tasks are returned: 2 for each environment

Reply:
Yeah I'll fix that. I agree, is that just add 4 task tags and assert 4 exist.